### PR TITLE
(PDOC-23) Emit json

### PR DESCRIPTION
**Pull request by iankronquist**
Addresses [PDOC-23](https://tickets.puppetlabs.com/browse/PDOC-23).
Nick had some ideas about its current form and this will need to be extended once #46 is merged. I'm putting it up here anticipating our meeting Tuesday the 8th.

### CODE REVIEW

File: lib/puppet_x/puppetlabs/strings/yard/code_objects/method_object.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
+class YARD::CodeObjects::MethodObject
```
Comment (iankronquist) [inline lib/puppet_x/puppetlabs/strings/yard/code_objects/method_object.rb]: As opposed to patching Yard's `MethodObject` we could just create `Puppet4xFunctionObject` and `Puppet3xFunctionObject`s which inherit from it. There is some logic identifying puppet and ruby methods (`:method`) right now which would need to be modified as a result.

**Comment by iankronquist** [comment]
Here's a sample of what this emits right now:
```json
{
    "Puppet3xFunctions": {
        "code_object_type": "HostClassObject",
        "docstring": "",
        "parameters": null,
        "type_info": null
    },
    "Puppet4xFunctions": {
        "code_object_type": "HostClassObject",
        "docstring": "",
        "parameters": null,
        "type_info": null
    },
    "foo": {
        "code_object_type": "HostClassObject",
        "docstring": "",
        "parameters": [
            [
                "ident",
                "\"Bob\""
            ],
            [
                "age",
                "10"
            ]
        ],
        "type_info": [
            {
                "age": "Myint",
                "ident": "Mytype"
            }
        ]
    },

}

```

File: lib/puppet_x/puppetlabs/strings/yard/json_registry_store.rb [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+        schema[:defined_types] += obj.select do |key, val|
+          val.type == :definedtype
+        end.values
+        @serializer.serialize(schema.to_json)
+    end
+  end
+
+  # Override the serializer because it puts the data at a whacky path and, more
+  # importantly, mashals the data with a bunch of non-printable characters.
```
Comment (iankronquist) [inline lib/puppet_x/puppetlabs/strings/yard/json_registry_store.rb]: Should be 'marshals'

**Comment by iankronquist** [comment]
Rebased to bring it up to date.

**Comment by iankronquist** [comment]
What else does this PR need? I think I followed all Nick's advice on the output structure, but it might be nice if he took one more look at this. It also needs to be squashed.

**Comment by HAIL9000** [comment]
@iankronquist have you had a chance to look into these spec failures?

**Comment by iankronquist** [comment]
Oh dear, I didn't realize there were spec failures. I think I goofed some refactoring in 12f2af6. I'll take a look.

**Comment by iankronquist** [comment]
It looks like I made changes and forgot to update the name of the object I was interested in in the test. It should work now.

**Comment by HAIL9000** [comment]
Awesome, thanks @iankronquist! @nfagerlund @adreyer, if either of you have a chance to give this another once over that would be amazing. Trying to prevent this work from slipping through the cracks since we're getting so close

**Comment by nfagerlund** [comment]
I've got this installed and working, and am playing with it now. It works!

- It should be possible to get JSON on stdout (or a file of our choice) via a command line option or a separate action; having to generate the yard site in order to get the json isn't ideal for our use cases. 
- The `@example` values for 4.x functions aren't surviving into the JSON! This looks like a bug. 
- Still a little wobbly on the name of the `puppet_version` key for functions, but.... also not super fussed about it. Maybe `function_api_version` or something. But, eh!!! I won't hold up the PR on that!

I still like the format, I think we settled on something very usable. The remaining things holding me up from ripping out a bunch of garbage custom code in favor of Strings are mostly either bugs in the parsing that we'll identify over time, or shortcomings in the inputs I'm feeding it (gotta go through the types and use `@-directives` to deal with some of the un-parseable meta-programming, and might need some more kinds of directives to help do that).

**Comment by nfagerlund** [comment]
Ah, one more thing! We should have the JSON format documented somewhere, and ideally use a schema to enforce it! I think Alex and I will both want that before we can justify pushing this into production use, as a guarantee that it'll be less brittle than the garbage it's replacing.

**Comment by iankronquist** [comment]
@nfagerlund excellent

1. Regarding json to stdout, @HAIL9000 what is the status of strings command line flags? I know we currently pass flags to yard via `.yardopts` files. Is it reasonable to create a new set of command line flags for strings?
2. We completely ignore `@example` and all other tags right now. This should be easy to fix I think.
3. Trivial to fix. Again, if you have a better idea how to denote this, by all means let me know.
4. It's documented in the issue, but I think that description is out of date now. I'll go see if I can update the issue. @HAIL9000 where should I put the documentation for the schema? I assume a markdown file somewhere in the repo would be best.

**Comment by iankronquist** [comment]
@nfagerlund do you want the examples just for Puppet 4x functions, for 3x and 4x functions, or for all code?

**Comment by nfagerlund** [comment]
@iankronquist I'm really leaning towards "everything." If there's a convention of not using it, like in the 3.x functions, that'll stay fine, but I can see people wanting to use these in classes, custom resource types, etc. Let's make it a first class citizen everywhere. 

Also: If you use an example in, say, a parameter docstring (for a resource type), it should probably attach to the parameter rather than the parent resource type.

**Comment by iankronquist** [comment]
@nfagerlund Done.

**Comment by iankronquist** [comment]
Okay I have now implemented two command line flags:
* `--emit-json $FILE`
* `--emit-json-stdout`

**Comment by nfagerlund** [comment]
@iankronquist Sweet. I've found three small problems:

* Both options are undocumented. I can't remember how to do that with a face, but I remember it was pretty simple.
* `--emit-json` can't seem to write to files that aren't in the cwd. I tired `~/Desktop/stdlib.json` and `/Users/nick/Desktop/stdlib.json`, and neither worked. 
* `--emit-json-stdout` adds some junk after the JSON data: 

```
Files:         115
Modules:         5 (    3 undocumented)
Classes:         1 (    0 undocumented)
Constants:       0 (    0 undocumented)
Methods:       123 (   14 undocumented)
Puppet Classes:     2 (    0 undocumented)
Puppet Defined Types:     0 (    0 undocumented)
Puppet Types:     2 (    2 undocumented)
Puppet Providers:     1 (    1 undocumented)
 85.07% documented
```

**Comment by iankronquist** [comment]
@nfagerlund All fixed.

**Comment by nfagerlund** [comment]
@iankronquist Rockin. Just tried it and it looks good. This could very well be good to go. @HAIL9000?

File: lib/puppet_x/puppetlabs/strings/yard/code_objects/method_object.rb (hunk 1 above)
Comment (HAIL9000) [inline lib/puppet_x/puppetlabs/strings/yard/code_objects/method_object.rb]: Would you mind filing a ticket for this improvement so we don't lose track of it?

**Comment by HAIL9000** [comment]
This all looks :+1: to me! @iankronquist would you mind updating the README to reflect this new functionality (including a description of the flags and how they can be used, and probably a reference to the JSON schema) and then squashing down to fewer commits? Then I think we can merge this.

**Comment by iankronquist** [comment]
@HAIL9000 Updated README and squashed down to 4 commits.

File: lib/puppet_x/puppetlabs/strings/yard/code_objects/method_object.rb (hunk 1 above)
Comment (iankronquist) [inline lib/puppet_x/puppetlabs/strings/yard/code_objects/method_object.rb]: [PDOC-64](https://tickets.puppetlabs.com/browse/PDOC-64) has been filed.

**Comment by iankronquist** [comment]
Yay! Thanks @nfagerlund and @HAIL9000 for your help finishing this up.

**Pull request merged**